lib/types/src/progress.rs

```diff
@@ -0,0 +1,107 @@
+//! Types used to report and handle compilation progress.
+
+use crate::lib::std::{borrow::Cow, fmt, string::String, sync::Arc};
+
+/// Indicates the current compilation progress.
+///
+/// All fields are kept private for forwards compatibility and future extension.
+/// Use the provided methods to access progress data.
+#[derive(Clone, Debug, Default)]
+pub struct CompilationProgress {
+    phase_name: Option<Cow<'static, str>>,
+    phase_step_count: Option<u64>,
+    phase_step: Option<u64>,
+}
+
+impl CompilationProgress {
+    /// Creates a new [`CompilationProgress`].
+    pub fn new(
+        phase_name: Option<Cow<'static, str>>,
+        phase_step_count: Option<u64>,
+        phase_step: Option<u64>,
+    ) -> Self {
+        Self {
+            phase_name,
+            phase_step_count,
+            phase_step,
+        }
+    }
+
+    /// Returns the name of the phase currently being executed.
+    pub fn phase_name(&self) -> Option<&str> {
+        self.phase_name.as_deref()
+    }
+
+    /// Returns the total number of steps in the current phase, if known.
+    pub fn phase_step_count(&self) -> Option<u64> {
+        self.phase_step_count
+    }
+
+    /// Returns the index of the current step within the phase, if known.
+    pub fn phase_step(&self) -> Option<u64> {
+        self.phase_step
+    }
+}
+
+/// Error returned when the user requests to abort an expensive computation.
+#[derive(Clone, Debug)]
+pub struct UserAbort {
```

Comment:
What about making it use `thiserror` as I see you implement `Display` trait anyway.